Create `v4/products/` endpoint

gigitux opened the pull request:
> [!NOTE]
> This PR is blocked by #60634  

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:


Part of https://github.com/woocommerce/woocommerce/issues/60704 (WOOPLUG-5461)

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR creates the endpoint `/products` under the `v4`. Currently the implementation is the same as the `v3`. I just copied-pasted `plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php` and `plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller-tests.php`.

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

For more details, see:  p1756738321043929-slack-C08UYCAP0JU
### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|        |       |


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Get this PR locally.
2. Enable `rest-api-v4` experimental feature via the WC Beta tester plugin.
3. Go to Woo > Settings > Advanced > REST API and create a new key.
4. In your preferred app, or using cURL: `GET /wp-json/wc/v4/products` with Basic Auth using the credentials created in the prior step.
5. Try `GET /wp-json/wc/v4/products` to retrieve a list of all products available in the store.
 
<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Create v4/products/ endpoint.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

elazzabi committed:
Add a new v4 controller

elazzabi committed:
Add a simple ping route

elazzabi committed:
Add get_v4_controllers and wc/v4 to the controllers list

elazzabi committed:
Add wc/v4 to should load namespace

elazzabi committed:
Add testing for the ping route

elazzabi committed:
Load v4 subdirectories loading

elazzabi committed:
add wc_rest_api_v4_is_enabled function

elazzabi committed:
Add the v4 to the list of Features

elazzabi committed:
Only load v4 if feature is enabled

elazzabi committed:
Load V4 and subdirectories only if feature is enabled

elazzabi committed:
Return success from recursive loader to allow autoloader to short-circuit.

elazzabi committed:
Call instance method on the REST server (fix static call).

elazzabi committed:
Remove problematic current_time

elazzabi committed:
Harden feature-flag helper to avoid fatal when container isn’t ready.

elazzabi committed:
declare strict types and add tearDown

elazzabi committed:
remove timestamp from testing cases

elazzabi committed:
Fix linting issues

elazzabi committed:
Remove unnecessary test

elazzabi committed:
Fix assert array to 2 instead of 3

elazzabi committed:
remove unnecessary declaration

elazzabi committed:
remove test file for now

elazzabi committed:
Merge branch 'trunk' into add/support-for-v4-api

elazzabi committed:
Force GH actions

elazzabi committed:
Remove the experimental feature from the UI

gigitux committed:
copy paste wc/v3 products endpoint implementation

gigitux committed:
create v4 endpoint for /products

elazzabi committed:
Rename feature to match other features (using - instead of _)

elazzabi committed:
Add rest-api-v4 to the development.json config

elazzabi committed:
Merge branch 'trunk' into add/support-for-v4-api

gigitux committed:
use right Features class

gigitux committed:
Merge branch 'add/support-for-v4-api' of github.com:woocommerce/woocommerce into add/v4-products-endpoint

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into add/v4-products-endpoint

gigitux committed:
Merge branch 'add/v4-products-endpoint' of github.com:woocommerce/woocommerce into add/v4-products-endpoint

gigitux committed:
add unit test

gigitux committed:
add unit test

gigitux committed:
ensure that unit test runs correctly

gigitux committed:
Merge branch 'add/v4-products-endpoint' of github.com:woocommerce/woocommerce into add/v4-products-endpoint

gigitux committed:
remove wrong folder

gigitux committed:
add README.md file

gigitux committed:
fix enable/disable feature flag on unit test

gigitux commented on plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version4/Products/class-wc-rest-products-v4-controller-tests.php:
```diff
@@ -0,0 +1,1994 @@
+		$this->disable_rest_api_v4_feature();
+	}
+
+	/**
+	 * Enable the REST API v4 feature.
+	 */
+	public static function enable_rest_api_v4_feature() {
+		add_filter(
+			'woocommerce_admin_features',
+			function ( $features ) {
+				$features[] = 'rest-api-v4';
+				return $features;
+			},
+		);
+	}
+
+	/**
+	 * Disable the REST API v4 feature.
+	 */
+	public static function disable_rest_api_v4_feature() {
+		add_filter(
+			'woocommerce_admin_features',
+			function ( $features ) {
+				$features = array_diff( $features, array( 'rest-api-v4' ) );
+				return $features;
+			}
+		);
+	}
```
Copied pasted from `plugins/woocommerce/tests/php/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller-tests.php`. I added only `enable_rest_api_v4_feature` and `disable_rest_api_v4_feature` to ensure that the feature flag is enabled before running unit tests.

gigitux committed:
Merge branch 'trunk' into add/v4-products-endpoint

gigitux commented on plugins/woocommerce/includes/wc-rest-functions.php:
```diff
@@ -466,13 +466,5 @@ function wc_rest_should_load_namespace( string $ns, string $rest_route = '' ): b
  * @return bool True if the REST API v4 feature is enabled, false otherwise.
  */
 function wc_rest_api_v4_is_enabled(): bool {
-	// Container may not be initialized yet during early autoloading.
-	if ( ! function_exists( 'wc_get_container' ) || empty( $GLOBALS['wc_container'] ) ) {
-		return false;
-	}
-	$container = wc_get_container();
-	if ( ! is_object( $container ) ) {
-		return false;
-	}
```
This logic isn't necessary. I'm going to remove it. cc @elazzabi

dinhtungdu commented on plugins/woocommerce/includes/rest-api/Controllers/Version4/Products/README.md:
```diff
@@ -0,0 +1,45 @@
+## Architecture
+
+- **Namespace**: `wc/v4`
+- **Future Goal**: Eventually migrate to extend `WP_REST_Controller` directly instead of WooCommerce base classes
```
I think we should list implementing the new version under the WooCommerce namespace as a goal, too.

gigitux replied:
Can you clarify this point? 🤔

dinhtungdu replied:
I mean moving the API codebase under the `src` directory and namespacing it under `Automattic\WooCommerce`.

dinhtungdu replied:
Same as what @mikejolley suggested.

gigitux replied:
Addressed with 1669ac74f6334b5da272937c55cd4cbd3ed62947

dinhtungdu commented on plugins/woocommerce/includes/wc-rest-functions.php:
```diff
@@ -466,13 +466,5 @@ function wc_rest_should_load_namespace( string $ns, string $rest_route = '' ): b
  * @return bool True if the REST API v4 feature is enabled, false otherwise.
  */
 function wc_rest_api_v4_is_enabled(): bool {
```
At this point, can we remove this global function?

gigitux replied:
I removed it 4c0172eccc9833445596007a1131ee7973ff58cb

dinhtungdu reviewed (commented):
<img width="1505" height="850" alt="Image" src="https://github.com/user-attachments/assets/86241ba6-71bb-4801-9f8b-f81a59e55d88" />

The route works as expected in my test. I only have some questions.

elazzabi committed:
Remove rest api v4 from FeaturesController

elazzabi committed:
Use Features::is_enabled directly

elazzabi committed:
Use Features::is_enabled and remove extra functions

gigitux committed:
remove is_v4_feature_enabled function

gigitux committed:
Merge branch 'add/v4-products-endpoint' of github.com:woocommerce/woocommerce into add/v4-products-endpoint

elazzabi committed:
Remove extra line

gigitux committed:
Merge branch 'fix/v4-api-experimental' of github.com:woocommerce/woocommerce into add/v4-products-endpoint

dinhtungdu reviewed (approved):
Thank you for the update. This PR is LGTM now. I noticed some phpcs errors and think we should address them. There're warnings about slow queries; however, it shouldn't prevent us from marketing this PR. I’m approving it for now.

gigitux committed:
Revert "Merge branch 'fix/v4-api-experimental' of github.com:woocommerce/woocommerce into add/v4-products-endpoint"

This reverts commit be1f43028170d577f80dbdf6f2b1cc60f6362861, reversing
changes made to 8c216f98131a41ab9549f2e3904b76a748d7d228.

gigitux committed:
fix lint errors

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into add/v4-products-endpoint

gigitux committed:
remove not related changes

gigitux committed:
update documentation

gigitux committed:
fix lint error

gigitux commented:
I reverted some changes since they will be shipped with #60708. I also fixed the linter errors and disabled the rule for that specific line to address the warning.
